### PR TITLE
fix: detect export destructring in tree-shaking

### PR DESCRIPTION
**Pull request by HerringtonDarkholme**
## Summary

ECMA spec supports export binding with destructring. This pull request supports this pattern, which is widely used in redux.

## Related issue (if exists)

#2521 

## Types of changes

- [ ] Docs change / Dependency upgrade
- [x] Bug fix
- [ ] New feature / Improvement
- [ ] Refactoring
- [ ] Breaking change

## Checklist

- [x] I have added changeset via `pnpm run changeset`. 
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: crates/rspack_core/src/tree_shaking/optimizer.rs [hunk 1]
```diff
@@ -554,6 +554,7 @@ impl<'a> CodeSizeOptimizer<'a> {
     }
     // visited_module.remove(&cur);
 
+    // TODO: this code can be slightly more concise
```
Comment (IWANABETHATGUY) [inline crates/rspack_core/src/tree_shaking/optimizer.rs:557]: I am not sure if you mean to use a boolean instead, the reason why I use a list here is `Vec` is debug-friendly. If not, could you provide more details?

File: crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+const foo = {
+  noop: () => {},
+	rename: 114,
+	otherFields: 514,
+};
+
+export const { noop, rename: reassign, ...rest } = foo;
```
Comment (hyf0) [inline crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js:7]: I'm not sure it works. 

Even though, we could analyze which variable is not used in `{ noop, rename: reassign, ...rest } = foo`.

But the tree shaker(which is the SWC minfier) we rely on doesn't shake unused variables in pattern like `{ noop, rename: reassign, ...rest } = foo`.

See https://play.swc.rs/?version=1.3.44&code=H4sIAAAAAAAAA0vOzysuUUjLz1ewVajmUlDIy88vsFLQ0FSwtVOortXh4ixKzUvMTbVSMDQ0AfLySzJSi9wyU3NSiq0UTEFCtdZcyWAzqsF6dRQgGnQUkJQq1AJNB9oBVpmfk6qXk5%2BuAVKtCQCVeh%2FZfwAAAA%3D%3D&config=H4sIAAAAAAAAA31VuY7jMAzt5ysC1VssUmwx7Q7S7TcIikQ5nLVFQ6QyCQb596GP3HK6WI%2BPN1%2B%2B31Yr88nevK%2B%2B9ad%2B9C4z5Mu3vvAxiTvoi5FjD%2Bwz9mJ%2BndEAnrITyqwWkgtcEOHD%2FDS%2BnCbASHaJI%2BXuNkZh%2BICICTaU%2F7aOeYPQhieXGZyXG97wVJJgB0N6rgh1TtCbGT7dB3a5gYFtgNe%2F1%2Bu5BtMS8cCPruU5lOkwYTzeZuip6zMw30dXl6WDJHzPn7FMX48l6PuWqAXtwTJiHVtMAg3kmmNPbet6Brt3T12fM3UZmWohBrAIBNtn6qt4CihISWM%2BowFcsJ4CVCDM4AX3UKNpLKUl1vIq9YxwgG1pmnHzHtiwd21xUokJh3Ekmm3F646QxcaSai2cwIUeTODc3EcmRptBSk7PvE%2FCtDCT%2FwDagWGxk%2Bug5ne0iLpPS%2Bz4kolJrwflWMF1v2tVJmi0qRYxVjo7dAayYG2aGULxMHTW19KZ4YX2MQawEKPuSsU1f6H4XS3oID0UK4DO18XaVk2AvVzhAj4cxAt4o1VKfcFmC1Wc3TLKx25L7YsAHciOwgsDHYXQMpxVJQ79Ml5SGIU1VE1Ud0NVBPQAhGw76uXTbuh5qEfbtLS9ysRZcy9i3bnUXO99kuK32cB0FMoIzn87w3wnaf5jrkZnFb4kbpD%2FnZlj0NMPG2ZSgsIGAAA%3D

File: crates/rspack_core/src/tree_shaking/optimizer.rs (hunk 1 above)
Comment (HerringtonDarkholme) [reply crates/rspack_core/src/tree_shaking/optimizer.rs:557]: We can use 

```
match side_effects_map.get_mut() {
  Some(SideEffect::Configuration(_)) |   None => vec![],
  Some(SideEffect::Analyze(_)) => {
    xxx
  }
}
```

Comment (IWANABETHATGUY) [reply crates/rspack_core/src/tree_shaking/optimizer.rs:557]: LGTM

File: crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js (hunk 2 above)
Comment (HerringtonDarkholme) [reply crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js:7]: The visitor change here only involves export symbols. Is swc-minifier's optimizing behavior affecting the visitor? 

What if our analysis has different output with swc-minifiers?

Comment (hyf0) [reply crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js:7]: The process is

1. We analyze which variable is unused.
2. Remove the `export` keyword for unused variables
3. perform SWC minifier to shake unused variables/AST nodes

The problem is we successfully remove the `export` keyword for unused variables but the SWC minifier doesn't support to shake code pattern mentioned above.

File: crates/rspack_core/src/tree_shaking/optimizer.rs (hunk 1 above)
Comment (HerringtonDarkholme) [reply crates/rspack_core/src/tree_shaking/optimizer.rs:557]: I can make the change in following pr. Keep this change relevant only to tree-shaking

File: crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js (hunk 2 above)
Comment (HerringtonDarkholme) [reply crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js:7]: Gotcha. Where is the remove export keyword code? The problem here is that we cannot remove export for cases like this 

```
export const { used, notUsed } = {...}
```

This is a partial used binding, changing it to `const { ...} = {...}` is not correct.

Comment (hyf0) [reply crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js:7]: Yeah. That's the problem for patterns like `export const { used, notUsed } = {...}`. 

I have investigated such cases before. That's why I'm not sure this PR works.

Comment (HerringtonDarkholme) [reply crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js:7]: Indeed tree shaking will not work on export destructring.

image
image

Comment (hyf0) [reply crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js:7]: If you have an interest in tree-shaking, you could try to investigate this case.

@IWANABETHATGUY Have we supported tree-shaking on cases like this?

```js
// a.js
export const unused = 1, used = 1;

// index.js
import { used } from './a.js'
console.log(used)
```

=> transform

```js
// a.js
const unused = 1;
export const used = 1;

// index.js
import { used } from './a.js'
console.log(used)
```

=> finally

```js
// a.js
export const used = 1;

// index.js
import { used } from './a.js'
console.log(used)
```

Comment (IWANABETHATGUY) [reply crates/rspack_error/tests/fixtures/should_not_warn_export_destructring/a.js:7]: yes, we have support this case

**Comment by IWANABETHATGUY** [comment]
here is the webpack behavior: 
## index.js
```js
import {a, } from './app'
a;

```
## app.js
```js
const test = {
  a: "bbbb",
  b: "aaaaa"
}

export const {a, b} = test;
```

## output
```js
(() => {
	"use strict";
	var a = {
			"./basic3/app.js": (a, r, e) => {
				e.d(r, { a: () => t });
				const { a: t, b: o } = { a: "bbbb", b: "aaaaa" };
			},
		},
		r = {};
	function e(t) {
		var o = r[t];
		if (void 0 !== o) return o.exports;
		var p = (r[t] = { exports: {} });
		return a[t](p, p.exports, e), p.exports;
	}
	(e.d = (a, r) => {
		for (var t in r)
			e.o(r, t) &&
				!e.o(a, t) &&
				Object.defineProperty(a, t, { enumerable: !0, get: r[t] });
	}),
		(e.o = (a, r) => Object.prototype.hasOwnProperty.call(a, r)),
		e("./basic3/app.js").a;
})();

```
`webpack` or `terser` did not tree shake destructor either.

**Comment by HerringtonDarkholme** [comment]
Hi @hyf0, I think this change is good to go now. Though it does not implement tree-shaking named-export, for good reason, at the moment, the PR suppresses bogus warnings on named export. Besides it, the change does not introduce observable change to the generated code.

**Review by hyf0 — APPROVED** [review]

**Pull request merged**